qcodes/dataset/sqlite_base.py

```diff
@@ -209,10 +232,86 @@ def connect(name: str, debug: bool = False) -> sqlite3.Connection:
 
     if debug:
         conn.set_trace_callback(print)
+
+    init_db(conn)
+    perform_db_upgrade(conn, version=version)
     return conn
 
 
-def transaction(conn: sqlite3.Connection,
+def perform_db_upgrade(conn: SomeConnection, version: int=-1) -> None:
+    """
+    This is intended to perform all upgrades as needed to bring the
+    db from version 0 to the most current version (or the version specified).
+    All the perform_db_upgrade_X_to_Y functions most succeed or raise.
+
+    Args:
+        version: Which version to upgrade to. We count from 0. -1 means
+          'latest version'
+    """
+
+    upgrade_actions = {1: (perform_db_upgrade_0_to_1,),
+                       -1: (perform_db_upgrade_0_to_1,)}
+
+    current_version = get_user_version(conn)
+    if current_version < 1 and (version > 0 or version == -1):
```

Comment:
This does not seem very general and directly extendable. Perhaps something like this would be simpler

```
upgrade_actions = {1: (perform_db_upgrade_0_to_1,)}
max_target_version = 1
if version == -1:
    version = max_target_version 

current_version = get_user_version(conn)
if current_version < version:
        log.info("Commencing database upgrade")
        for action in upgrade_actions[version]:
            action(conn)

```

Reply:
Sold!